qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,306 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from base64 import urlsafe_b64decode
+from string import ascii_letters, digits
+import datetime
+from random import SystemRandom
+import functools
+from traceback import format_exception
+
+from tornado.web import RequestHandler
+
+from moi import r_client
+from qiita_core.exceptions import (IncorrectPasswordError, IncorrectEmailError,
+                                   UnverifiedEmailError)
+import qiita_db as qdb
+
+
+def login_client(client_id, client_secret=None):
+    """Login client_id, optionally with client_secret
+
+    Parameters
+    ----------
+    client_id : str
+        The client ID to validate
+    client_secret : str, optional
```

Comment:
I had to read the entire code to understand why this is correct. My take on this is that this is really dangerous and it is really easy to introduce bugs in the code.

Now, from the code structure correctness, this function is used in 2 places, the function `validate_resource_owner` and the function `validate_client`. In this function, you have an if statement, which the only thing that is doing is differentiating the 2 cases that is calling this function. This exposes that there is no functionality in common and this function should've not been decomposed in a single function - I'm not even sure that is saving lines of code...

Can you remove this function and just write the call in the given functions?